### ISSUE TITLE
Multiline input hangs logstash

### ISSUE THREAD

**Issue by runningman84** [elasticsearch/logstash#1481]
We are using logstash (version 1.4.1 but 1.3.3 is suffering from the same problem) to read error logfiles, parse them and forward them to graylog2 using gelf. Everything works fine until we use the multiline codec as input or as filter.

This leads to 100% cpu usuage:
![logstash_top](https://cloud.githubusercontent.com/assets/1699128/3412992/87dc32b6-fe06-11e3-9657-68621e6ea8c6.PNG)

At this point graylog2 doesn't receive any futher logfiles. Restarting logstash temporarily solves this issue but eventually it will hang again.

These are some recent thread dumps collected while logstash was hanging.
http://pastebin.com/GYSCG2z2
http://pastebin.com/zn622qgQ

This problem happens after a while, sometimes logstash runs for a whole day, sometimes logstash hangs after a few minutes. We read files which contain java error messages and stack straces. These files grow up to 3GB per day.

We use inputs like this:
```
        file {
                path =>  "/EXAMPLE/logs/slowrequests.log"
                codec => multiline {
                   pattern => "^[0-9]+\.[0-3][0-9]\.20[0-3][0-9] "
                   negate => true
                   what => previous
                } 
        }
```

A sample log entry looks like this:
```
27.06.2014 00:01:39.564 *WARN* [RMI TCP Connection(899)-127.0.0.1] org.apache.jackrabbit.core.query.lucene.NodeIteratorImpl Failed to retrieve query result node f6c40ccb-3215-4d28-aafe-be8188f3de54 javax.jcr.ItemNotFoundException: f6c40ccb-3215-4d28-aafe-be8188f3de54
        at org.apache.jackrabbit.core.ItemManager.getItemData(ItemManager.java:384)
        at org.apache.jackrabbit.core.ItemManager.getItem(ItemManager.java:328)
        at org.apache.jackrabbit.core.ItemManager.getItem(ItemManager.java:622)
        at org.apache.jackrabbit.core.query.lucene.NodeIteratorImpl$FetchNext.perform(NodeIteratorImpl.java:197)
        at org.apache.jackrabbit.core.session.SessionState.perform(SessionState.java:216)
        at org.apache.jackrabbit.core.query.lucene.NodeIteratorImpl.fetchNext(NodeIteratorImpl.java:183)
        at org.apache.jackrabbit.core.query.lucene.NodeIteratorImpl.initialize(NodeIteratorImpl.java:214)
        at org.apache.jackrabbit.core.query.lucene.NodeIteratorImpl.hasNext(NodeIteratorImpl.java:163)
        at org.apache.sling.jcr.resource.internal.helper.jcr.JcrNodeResourceIterator.seek(JcrNodeResourceIterator.java:90)
        at org.apache.sling.jcr.resource.internal.helper.jcr.JcrNodeResourceIterator.<init>(JcrNodeResourceIterator.java:64)
        at org.apache.sling.jcr.resource.internal.helper.jcr.JcrResourceProvider.findResources(JcrResourceProvider.java:263)
        at org.apache.sling.resourceresolver.impl.tree.RootResourceProviderEntry$2.seek(RootResourceProviderEntry.java:144)
        at org.apache.sling.resourceresolver.impl.tree.RootResourceProviderEntry$2.<init>(RootResourceProviderEntry.java:134)
```

**Comment by untergeek** [comment]
I have used your data and followed the steps outlined with the current master branch (pre-Logstash 1.5) and the current versions of all plugins mentioned in your config and cannot reproduce the problem.  As a result, I'm going to suggest that this has been addressed by Logstash 1.5.

If there are other concerns, please open tickets at the respective plugins, namely:
https://github.com/logstash-plugins/logstash-input-file
https://github.com/logstash-plugins/logstash-codec-multiline
https://github.com/logstash-plugins/logstash-filter-grok
https://github.com/logstash-plugins/logstash-output-stdout

**Comment by colinsurprenant** [comment]
this is probably related to logstash-plugins/logstash-codec-multiline#4
